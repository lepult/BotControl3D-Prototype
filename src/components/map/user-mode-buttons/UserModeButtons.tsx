import React, { FC, useMemo } from 'react';
import '../buttons.scss';
import { useSelector } from 'react-redux';
import { getUser, useUser, useWindowMetrics } from 'chayns-api';
import { selectMapIds } from '../../../redux-modules/map/selectors';
import FloorSelectionButton from './floor-buttons/FloorSelectionButton';
import { selectRobotIds } from '../../../redux-modules/robot-status/selectors';
import RobotSelectionButton from './robot-selection-buttons/RobotSelectionButton';
import RouteButton from './robot-controls-buttons/route/RouteButton';
import ChargeButton from './robot-controls-buttons/charge/ChargeButton';
import CancelButton from './robot-controls-buttons/cancel/CancelButton';
import FollowRobotButton from './robot-controls-buttons/follow-robot/FollowRobotButton';
import ResetViewButton from './interaction-buttons/ResetViewButton';
import { selectIsPlanningRoute } from '../../../redux-modules/misc/selectors';
import { getUserType } from '../../../utils/permissionsHelper';
import { UserType } from '../../../types/misc';
import InformationButton from './robot-controls-buttons/information/InformationButton';
import EditMapButton from './interaction-buttons/EditMapButton';

const UserModeButtons: FC = () => {
    const allMapIds = useSelector(selectMapIds);
    const allRobotIds = useSelector(selectRobotIds);
    const metrics = useWindowMetrics();
    const isPlanningRoute = useSelector(selectIsPlanningRoute);

    const user = useUser();
    const userType = useMemo(() => getUserType(user), [user]);
    const isGuest = useMemo(() => userType === UserType.guest, [userType]);

    return (
        <div className="user-mode-buttons__wrapper">
            <div
                style={{
                    position: 'relative',
                    width: '100%',
                    height: '100%',
                }}
            >
                {(metrics.pageWidth > 900 || !isPlanningRoute) && (
                    <div
                        className="map-buttons map-buttons-row position-left position-top bot-selection-buttons"
                        style={{
                            flexWrap: 'wrap',
                        }}
                    >
                        {allRobotIds.map((id) => (
                            <RobotSelectionButton
                                key={id}
                                robotId={id as string}
                            />
                        ))}
                    </div>
                )}

                <div
                    className="map-buttons"
                    style={{
                        alignItems: 'end',
                        top: 0,
                        left: 0,
                        right: 0,
                        padding: '10px',
                    }}
                >
                    {!isGuest && <RouteButton/>}
                    {!isPlanningRoute && !isGuest && <CancelButton/>}
                    {!isPlanningRoute && !isGuest && <ChargeButton/>}
                </div>
                <div className="map-buttons position-left position-bottom">
                    {allMapIds.map((id) => (
                        <FloorSelectionButton
                            key={id}
                            mapId={id}
                        />
                    ))}
                </div>
                <div className="map-buttons position-right position-bottom">
                    <FollowRobotButton/>
                    <InformationButton/>
                    {user.uacGroups?.find((group) => group.id === 1) && <EditMapButton/>}
                    <ResetViewButton/>
                </div>
            </div>
        </div>
    )
};

export default UserModeButtons;
