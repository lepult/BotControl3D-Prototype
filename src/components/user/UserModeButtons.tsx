import React, { FC } from 'react';
import './userModeButtons.scss';
import { useDispatch, useSelector } from 'react-redux';
// @ts-ignore
import { Button } from 'chayns-components';
import { selectFollowRobot, selectMapIds, selectSelectedRobot } from '../../redux-modules/map/selectors';
import FloorButton from './FloorButton';
import { selectRobotIds } from '../../redux-modules/robot-status/selectors';
import RobotButton from './RobotButton';
import { toggleFollowRobot } from '../../redux-modules/map/actions';

const UserModeButtons: FC = () => {
    const dispatch = useDispatch();

    const allMapIds = useSelector(selectMapIds);
    const allRobotIds = useSelector(selectRobotIds);
    const selectedRobot = useSelector(selectSelectedRobot);
    const followRobot = useSelector(selectFollowRobot);

    return (
        <div className="user-mode-buttons__wrapper">
            <div
                className="map-buttons map-buttons-row position-left position-top"
                style={{ flexWrap: 'wrap' }}
            >
                {allRobotIds.map((id) => (
                    <RobotButton robotId={id as string}/>
                ))}
            </div>
            <div className="map-buttons position-right position-top">
                <Button
                    disabled={!selectedRobot}
                    className={followRobot ? '' : 'button--secondary'}
                    onClick={() => dispatch(toggleFollowRobot())}
                >
                    <i className="fa-solid fa-location-arrow"/>
                </Button>
            </div>
            <div className="map-buttons position-left position-bottom">
                {allMapIds.map((id) => (
                    <FloorButton mapId={id}/>
                ))}
            </div>
            <div className="map-buttons position-right position-bottom">
                Buttons Unten Rechts
            </div>
        </div>
    )
};

export default UserModeButtons;
