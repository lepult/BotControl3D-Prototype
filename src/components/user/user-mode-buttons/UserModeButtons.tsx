import React, { FC } from 'react';
import './userModeButtons.scss';
import { useSelector } from 'react-redux';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { selectMapIds } from '../../../redux-modules/map/selectors';
import FloorSelectionButton from './floor-buttons/FloorSelectionButton';
import { selectRobotIds } from '../../../redux-modules/robot-status/selectors';
import RobotSelectionButton from './robot-selection-buttons/RobotSelectionButton';
import RouteButton from './robot-controls-buttons/RouteButton';
import ChargeButton from './robot-controls-buttons/ChargeButton';
import CancelButton from './robot-controls-buttons/CancelButton';
import FollowRobotButton from './interaction-buttons/FollowRobotButton';

const UserModeButtons: FC = () => {

    const allMapIds = useSelector(selectMapIds);
    const allRobotIds = useSelector(selectRobotIds);

    return (
        <div className="user-mode-buttons__wrapper">
            <div
                className="map-buttons map-buttons-row position-left position-top"
                style={{ flexWrap: 'wrap' }}
            >
                {allRobotIds.map((id) => (
                    <RobotSelectionButton robotId={id as string}/>
                ))}
            </div>
            <div className="map-buttons position-right position-top">
                <RouteButton/>
                <ChargeButton/>
                <CancelButton/>
            </div>
            <div className="map-buttons position-left position-bottom">
                {allMapIds.map((id) => (
                    <FloorSelectionButton mapId={id}/>
                ))}
            </div>
            <div className="map-buttons position-right position-bottom">
                <FollowRobotButton/>
            </div>
        </div>
    )
};

export default UserModeButtons;
