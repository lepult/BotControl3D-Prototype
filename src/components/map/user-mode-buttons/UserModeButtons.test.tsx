/* eslint-disable @typescript-eslint/ban-ts-comment */
import React from 'react';
import renderer from 'react-test-renderer';
import UserModeButtons from './UserModeButtons';

const robotIds = [1, 2, 3];
const mapIds = [1, 2, 3];

const reduxState = {
    robotStatus: {
        ids: [...robotIds],
    },
    map: {
        ids: [...mapIds],
    },
    misc: {
        isPlanningRoute: false,
    }
}

const windowMetrics = { pageWidth: 1000 };

jest.mock('react-redux', () => ({
    useDispatch: () => console.log('DISPATCH!!!!!!'),
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-return
    useSelector: (selector) => selector(reduxState),
}));

jest.mock('chayns-api', () => ({
    useUser: () => ({
        uacGroups: [{ id: 1 }],
    }),
    useWindowMetrics: () => ({ ...windowMetrics })
}));

jest.mock('./robot-controls-buttons/route/RoutePlanner', () => () => <div className="route-planner"/>);
jest.mock('./robot-controls-buttons/route/RouteButton', () => () => <div className="route"/>);
jest.mock('./robot-controls-buttons/cancel/CancelButton', () => () => <div className="cancel"/>);
jest.mock('./robot-controls-buttons/charge/ChargeButton', () => () => <div className="charge"/>);
jest.mock('./robot-controls-buttons/follow-robot/FollowRobotButton', () => () => <div className="follow"/>);
jest.mock('./robot-controls-buttons/information/InformationButton', () => () => <div className="info"/>);
jest.mock('./interaction-buttons/EditMapButton', () => () => <div className="edit"/>);
jest.mock('./interaction-buttons/ResetViewButton', () => () => <div className="reset"/>);
jest.mock('./robot-selection-buttons/RobotSelectionButton', () => ({
    robotId
}: {
    robotId: number
}) => <div className="select-robot" id={`${robotId}`}/>);
jest.mock('./floor-buttons/FloorSelectionButton', () => ({
    mapId
}: {
    mapId: number
}) => <div className="select-floor" id={`${mapId}`}/>);

it('robotIds, mapIds and not planning route should result in all buttons showing', () => {
    const component = renderer.create(
        <UserModeButtons/>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('No robotIds should result in no robot buttons', () => {
    reduxState.robotStatus.ids = [];

    const component = renderer.create(
        <UserModeButtons/>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    reduxState.robotStatus.ids = [...robotIds];
});

it('No mapIds should result in no map buttons', () => {
    reduxState.map.ids = [];

    const component = renderer.create(
        <UserModeButtons/>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    reduxState.map.ids = [...mapIds];
});

it('isPlanningRoute should result in route planner showing', () => {
    reduxState.misc.isPlanningRoute = true;

    const component = renderer.create(
        <UserModeButtons/>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    windowMetrics.pageWidth = 100;

    component.update(<UserModeButtons/>);
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    windowMetrics.pageWidth = 1000;
    reduxState.misc.isPlanningRoute = false;
});