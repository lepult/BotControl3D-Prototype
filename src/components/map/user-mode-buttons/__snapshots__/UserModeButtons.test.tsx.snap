// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`No mapIds should result in no map buttons 1`] = `
<div
  className="user-mode-buttons__wrapper"
>
  <div
    style={
      {
        "height": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      className="map-buttons map-buttons-row position-left position-top bot-selection-buttons"
      style={
        {
          "flexWrap": "wrap",
        }
      }
    >
      <button
        className="select-robot"
        id="1"
      />
      <button
        className="select-robot"
        id="2"
      />
      <button
        className="select-robot"
        id="3"
      />
    </div>
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "left": 0,
          "padding": "10px",
          "right": 0,
          "top": 0,
        }
      }
    >
      <button
        className="route"
      />
      <button
        className="cancel"
      />
      <button
        className="charge"
      />
    </div>
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "left": 0,
          "padding": "10px",
          "right": 0,
          "top": "50%",
          "transform": "translateY(-50%)",
        }
      }
    >
      <button
        className="follow"
      />
      <button
        className="info"
      />
    </div>
    <div
      className="map-buttons position-left position-bottom"
    />
    <div
      className="map-buttons position-right position-bottom"
    >
      <button
        className="edit"
      />
      <button
        className="reset"
      />
    </div>
  </div>
</div>
`;

exports[`No robotIds should result in no robot buttons 1`] = `
<div
  className="user-mode-buttons__wrapper"
>
  <div
    style={
      {
        "height": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      className="map-buttons map-buttons-row position-left position-top bot-selection-buttons"
      style={
        {
          "flexWrap": "wrap",
        }
      }
    />
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "left": 0,
          "padding": "10px",
          "right": 0,
          "top": 0,
        }
      }
    >
      <button
        className="route"
      />
      <button
        className="cancel"
      />
      <button
        className="charge"
      />
    </div>
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "left": 0,
          "padding": "10px",
          "right": 0,
          "top": "50%",
          "transform": "translateY(-50%)",
        }
      }
    >
      <button
        className="follow"
      />
      <button
        className="info"
      />
    </div>
    <div
      className="map-buttons position-left position-bottom"
    >
      <button
        className="select-floor"
        id="1"
      />
      <button
        className="select-floor"
        id="2"
      />
      <button
        className="select-floor"
        id="3"
      />
    </div>
    <div
      className="map-buttons position-right position-bottom"
    >
      <button
        className="edit"
      />
      <button
        className="reset"
      />
    </div>
  </div>
</div>
`;

exports[`isPlanningRoute should result in route planner showing 1`] = `
<div
  className="user-mode-buttons__wrapper"
>
  <div
    style={
      {
        "height": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      className="map-buttons map-buttons-row position-left position-top bot-selection-buttons"
      style={
        {
          "flexWrap": "wrap",
        }
      }
    >
      <button
        className="select-robot"
        id="1"
      />
      <button
        className="select-robot"
        id="2"
      />
      <button
        className="select-robot"
        id="3"
      />
    </div>
    <button
      className="route-planner"
    />
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "left": 0,
          "padding": "10px",
          "right": 0,
          "top": 0,
        }
      }
    >
      <button
        className="route"
      />
      <button
        className="cancel"
      />
      <button
        className="charge"
      />
    </div>
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "left": 0,
          "padding": "10px",
          "right": 0,
          "top": "50%",
          "transform": "translateY(-50%)",
        }
      }
    >
      <button
        className="follow"
      />
      <button
        className="info"
      />
    </div>
    <div
      className="map-buttons position-left position-bottom"
    >
      <button
        className="select-floor"
        id="1"
      />
      <button
        className="select-floor"
        id="2"
      />
      <button
        className="select-floor"
        id="3"
      />
    </div>
    <div
      className="map-buttons position-right position-bottom"
    >
      <button
        className="edit"
      />
      <button
        className="reset"
      />
    </div>
  </div>
</div>
`;

exports[`isPlanningRoute should result in route planner showing 2`] = `
<div
  className="user-mode-buttons__wrapper"
>
  <div
    style={
      {
        "height": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <button
      className="route-planner"
    />
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "left": 0,
          "padding": "10px",
          "right": 0,
          "top": "50%",
          "transform": "translateY(-50%)",
        }
      }
    >
      <button
        className="follow"
      />
      <button
        className="info"
      />
    </div>
    <div
      className="map-buttons position-left position-bottom"
    >
      <button
        className="select-floor"
        id="1"
      />
      <button
        className="select-floor"
        id="2"
      />
      <button
        className="select-floor"
        id="3"
      />
    </div>
    <div
      className="map-buttons position-right position-bottom"
    >
      <button
        className="edit"
      />
      <button
        className="reset"
      />
    </div>
  </div>
</div>
`;

exports[`robotIds, mapIds and not planning route should result in all buttons showing 1`] = `
<div
  className="user-mode-buttons__wrapper"
>
  <div
    style={
      {
        "height": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      className="map-buttons map-buttons-row position-left position-top bot-selection-buttons"
      style={
        {
          "flexWrap": "wrap",
        }
      }
    >
      <button
        className="select-robot"
        id="1"
      />
      <button
        className="select-robot"
        id="2"
      />
      <button
        className="select-robot"
        id="3"
      />
    </div>
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "left": 0,
          "padding": "10px",
          "right": 0,
          "top": 0,
        }
      }
    >
      <button
        className="route"
      />
      <button
        className="cancel"
      />
      <button
        className="charge"
      />
    </div>
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "left": 0,
          "padding": "10px",
          "right": 0,
          "top": "50%",
          "transform": "translateY(-50%)",
        }
      }
    >
      <button
        className="follow"
      />
      <button
        className="info"
      />
    </div>
    <div
      className="map-buttons position-left position-bottom"
    >
      <button
        className="select-floor"
        id="1"
      />
      <button
        className="select-floor"
        id="2"
      />
      <button
        className="select-floor"
        id="3"
      />
    </div>
    <div
      className="map-buttons position-right position-bottom"
    >
      <button
        className="edit"
      />
      <button
        className="reset"
      />
    </div>
  </div>
</div>
`;
